### PR TITLE
Implemented various graph utility methods.

### PR DESCRIPTION
**Pull request by zaunerc**
We use the JGraphT library in one of our software projects. We thought a few utility methods for common use cases were missing and implemented them ourselves. A quick Google search confirms that other projects would also benefit if the JGraphT library would implement graph utility methods for such common use cases. See e.g. https://github.com/Netflix/Nicobar/blob/84074b5/nicobar-core/src/main/java/com/netflix/nicobar/core/module/GraphUtils.java.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 1]
```diff
@@ -433,6 +435,228 @@
...
+    protected static <V, E extends DefaultEdge> List<V> determineDescendants(
+        DirectedAcyclicGraph<V, E> graph,
+        V vertex,
+        List<V> descendants)
+    {
+
+        List<V> children = Graphs.successorListOf(graph, vertex);
+
+        for (V child : children) {
+            descendants.add(child);
```
Comment (fulcorno) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: shouldn't be a check whether ```child``` is already in the list? Othewise, the same node might be added more than once (in the case of reconvergent forward paths)

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 2]
```diff
@@ -42,7 +42,9 @@
 package org.jgrapht;
 
 import java.util.*;
+import java.util.function.Predicate;
 
+import org.jgrapht.experimental.dag.DirectedAcyclicGraph;
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: We don't want dependencies from the main jgrapht packages on experimental.  So, the DAG-related methods should be moved to the DAG class itself (or a new companion utility class in the same package).

(Arguably, it's time to finally make DAG non-experimental, but we'll leave that for a separate effort.)

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 3]
```diff
@@ -433,6 +435,228 @@
...
+        for (V child : children) {
+            descendants.add(child);
+            determineDescendants(graph, child, descendants);
+        }
+
+        return descendants;
+    }
+
+    /**
+     * Removes the given vertex from the given graph. If the vertex to be
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Avoid "parent/child" terminology.  Graphs are not hierarchical.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 4]
```diff
@@ -433,6 +435,228 @@
...
+    /**
+     * This function is implemented recursively. Therefore only graphs without
+     * any cycles are allowed.
+     *
+     * @param graph graph to look for ancestors in.
+     * @param vertex the vertex to get the ancestors of.
+     *
+     * @return {@link List} of ancestors of the vertex in the given graph.
+     */
+    public static <V, E extends DefaultEdge> List<V> determineAncestors(
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Use "get" instead of "determine".

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 5]
```diff
@@ -433,6 +435,228 @@
...
+            }
+            graph.addEdge(source, target);
+        }
+    }
+
+    public static <V, E extends DefaultEdge> boolean vertexHasChildren(
+        DirectedGraph<V, E> graph,
+        V vertex)
+    {
+        return !Graphs.successorListOf(graph, vertex).isEmpty();
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Why use successorListOf, which is slower than the underlying outgoingEdgesOf?

File: jgrapht-core/src/test/java/org/jgrapht/GraphsTest.java [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+        DefaultDirectedGraph<String, TestEdge> graph = new DefaultDirectedGraph<String, TestEdge>(
+            TestEdge.class);
+
+        String a = "A";
+        String b = "B";
+        String c = "C";
+
+        graph.addVertex(b);
+
+        // System.out.println("Input:\t\t" + graph);
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/GraphsTest.java]: Delete all commented-out debugging code.

File: jgrapht-core/src/test/java/org/jgrapht/GraphsTest.java [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+import org.junit.runner.Description;
+
+/**
+ * @author Christoph Zauner
+ */
+public class GraphsTest
+{
+
+    // @Rule
+    public TestRule watcher = new TestWatcher()
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/GraphsTest.java]: This is unnecessary.

File: jgrapht-core/src/test/java/org/jgrapht/TestEdge.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * Original Author:  Christoph Zauner
+ *
+ * $Id$
+ *
+ * Changes
+ * -------
+ *
+ */
+package org.jgrapht;
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/TestEdge.java]: I guess this should live in org.jgrapht.graph instead (since that's where the DefaultEdge class is defined).

File: jgrapht-ext/pom.xml [hunk 9]
```diff
@@ -35,6 +35,19 @@
 					</execution>
 				</executions>
 				<configuration>
+					<transformers>
```
Comment (jsichi) [inline jgrapht-ext/pom.xml]: I don't think this change is related to the utility methods, so it should be in a separate pull request.

**Comment by zaunerc** [comment]
Thanks a lot for your constructive feedback! I will incorporate your suggested changes and update the PR.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 2 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: I moved the DAG-related methods to the DAG class itself (commit _9bd26b9c2ed9361dc291264a4fe2ee333fcccb1a_). There is no need for the new methods to be static anymore. Therefore I would like to convert the new methods in the DAG class to instance methods. What do you think?

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 4 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Done (see commit _455520a569acf8215314c5e2303a8a2e54ece823_).

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 1 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Thanks for the hint. You are absolutely right. I have added a check and corresponding unit tests. See commit _0385f60d7a389d2f50028d4b88e08705433182c8_.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 5 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Changed. See commit _3aa00a6018dafcb94c46e43a68924769aac6af9b_. As long as there is no such thing as a dangling edge ( `A --->` ) this approach seems fine ;-)

File: jgrapht-core/src/test/java/org/jgrapht/GraphsTest.java (hunk 6 above)
Comment (zaunerc) [inline jgrapht-core/src/test/java/org/jgrapht/GraphsTest.java]: Done (commit _07c6a8f6c6c428bf3745f0e39139b5dc1a3b9c66_).

File: jgrapht-core/src/test/java/org/jgrapht/GraphsTest.java (hunk 7 above)
Comment (zaunerc) [inline jgrapht-core/src/test/java/org/jgrapht/GraphsTest.java]: Done (commit _07c6a8f6c6c428bf3745f0e39139b5dc1a3b9c66_).

File: jgrapht-core/src/test/java/org/jgrapht/TestEdge.java (hunk 8 above)
Comment (zaunerc) [inline jgrapht-core/src/test/java/org/jgrapht/TestEdge.java]: Done (commit _43e5ef69a43aa687bee54100918e4ddbf301c999_).

File: jgrapht-ext/pom.xml (hunk 9 above)
Comment (zaunerc) [inline jgrapht-ext/pom.xml]: Done (commit _d52394d509b9043698c191d052bed1eaa91b90f8_). Will create a separate PR for this issue.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 3 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: What about the following terminology?

- successors
- direct successors
- predecessors
- direct predecessors

Therefore e.g.:

- `determineDescendants` -> `determineSuccessors`
- `removeVertexAndConnectParentsWithChildren` -> `removeVertexAndConnectDirectPredecessorsWithDirectSuccessors`
- `vertexHasChildren` -> `vertexHasDirectSuccessors`

**Comment by zaunerc** [comment]
I have attached a few comments and questions to your notes of my first commit (https://github.com/jgrapht/jgrapht/pull/197/commits/eefabb14b7547db6b9d46873c8a79792b1d454b5).

File: jgrapht-ext/pom.xml (hunk 9 above)
Comment (zaunerc) [inline jgrapht-ext/pom.xml]: Created #209.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 3 above)
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: We already use "successors" to mean "direct successors" and "predecessors" to mean "direct predecessors" (per https://en.wikipedia.org/wiki/Directed_graph).  We should keep doing so.

In the context of a DAG, the terms parent/child/ancestor/descendant are meaningful.  Luckily, since we decided to move the DAG-related methods to the DAG class, there will be no confusion if we keep using those.

So:

* keep determineDescendants as is (move to DAG)
* vertexHasChildren -> vertexHasSuccessors
* removeVertexAndConnectParentsWithChildren -> removeVertexAndPreserveConnectivity

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 2 above)
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: I agree.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 3 above)
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: (or rather, keep getDescendants as is)

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 10]
```diff
@@ -433,6 +435,159 @@
...
+     * removed has one or more parents, the parents will be connected directly
+     * to the children of the vertex to be removed.
+     *
+     * @param graph graph to be mutated.
+     * @param vertex vertex to be removed from this graph, if present.
+     *
+     * @return true if the graph contained the specified vertex; false
+     *         otherwise.
+     */
+    public static <V, E extends DefaultEdge> boolean removeVertexAndConnectParentsWithChildren(
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: why E extends DefaultEdge?

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 11]
```diff
@@ -433,6 +435,159 @@
...
+     * to the children of the vertex to be removed.
+     *
+     * @param graph to be mutated.
+     * @param vertices vertices to be removed from this graph, if present.
+     *
+     * @return true if at least one vertex has been removed; false otherwise.
+     */
+    public static <V, E extends DefaultEdge> boolean removeVerticesAndConnectParentsWithChildren(
+        DirectedGraph<V, E> graph,
+        List<V> vertices)
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Use Iterable instead of List.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java [hunk 12]
```diff
@@ -433,6 +435,159 @@
...
+    }
+
+    /**
+     * Add edges from one source vertex to multiple target vertices. Whether
+     * duplicates are created depends on the underlying {@link DirectedGraph}
+     * implementation.
+     *
+     * @param graph graph to be mutated
+     * @param source source vertex of the new edges
+     * @param targets target vertices for the new edges
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: use Iterable instead of List

File: jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java [hunk 13]
```diff
@@ -532,6 +533,82 @@ private void reorder(Set<V> df, Set<V> db, Visited visited)
     }
 
     /**
+     * This function is implemented recursively. Therefore only graphs without
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java]: (since this is on DAG now, you can delete the part of the comment about no cycles allowed)

File: jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java [hunk 14]
```diff
@@ -532,6 +533,82 @@ private void reorder(Set<V> df, Set<V> db, Visited visited)
...
+        DirectedAcyclicGraph<V, E> graph,
+        V vertex)
+    {
+
+        List<V> ancestors = new ArrayList<>();
+        ancestors = getAncestors(graph, vertex, ancestors);
+        return ancestors;
+    }
+
+    protected static <V, E extends DefaultEdge> List<V> getAncestors(
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java]: Since this is a recursive helper method, make it private and return void.

Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java]: Also, to avoid blowing the stack on very deep graphs, you could use DepthFirstIterator instead.

File: jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java [hunk 15]
```diff
@@ -532,6 +533,82 @@ private void reorder(Set<V> df, Set<V> db, Visited visited)
     }
 
     /**
+     * This function is implemented recursively. Therefore only graphs without
+     * any cycles are allowed.
+     *
+     * @param graph graph to look for ancestors in.
+     * @param vertex the vertex to get the ancestors of.
+     *
+     * @return {@link List} of ancestors of the vertex in the given graph.
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java]: since the ancestors are unique (and not guaranteed to be in any particular order), the return type should be Set, not List

File: jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java (hunk 14 above)
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java]: But if you do keep it recursive, then for the helper method, use "gather" instead of "get".

**Comment by jsichi** [comment]
I have responded to your previous comments, and added some new ones of my own.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 10 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: You are right. The utility methods do not depend on DefaultEdge in any way. Therefore this restriction is not required. I removed it. See commit 69bc6fca9f7be131c9afcc1b8fedb67916212877.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 11 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Done. See commit _860206d5b53451b2355b480a5334794f3fbf5e0c_.

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 12 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Done. See commit _860206d5b53451b2355b480a5334794f3fbf5e0c_.

File: jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java (hunk 14 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java]: I am now using iterators instead of recursion to determine the descandants/ancestors. If I would have known that determining the ancestors is as easy as reversing the graph using `EdgeReversedGraph` and iterating over it using the `DepthFirstIterator ` I would have done this in the first place :-) See commit _6d9a91c81918b2f05c8a6b8cdf1f6e41592b9d6b_.

File: jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java (hunk 15 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java]: Done (commit _6d9a91c81918b2f05c8a6b8cdf1f6e41592b9d6b_).

File: jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java (hunk 13 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/experimental/dag/DirectedAcyclicGraph.java]: Done (commit _6d9a91c81918b2f05c8a6b8cdf1f6e41592b9d6b_).

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 2 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: Done (commit _9c20641d06787d5802776c30a040f7628be353e3_).

File: jgrapht-core/src/main/java/org/jgrapht/Graphs.java (hunk 3 above)
Comment (zaunerc) [inline jgrapht-core/src/main/java/org/jgrapht/Graphs.java]: I have incorporated your suggestions. See commit _46610a6f415e83b0eb77122c73b4a56ca79bedec_.

**Comment by jsichi** [comment]
Looks good, merging now!

**Comment by zaunerc** [comment]
Thanks again for your constructive feedback and for merging the PR. Contributing to this project is a pleasure.
